unittests/savanna_proposer_policy_tests.cpp

```diff
@@ -0,0 +1,279 @@
+   // block produced, and we expect the new proposer policy to become active on the next first block of a round
+   // ---------------------------------------------------------------------------------------------------------
+   sb = A.produce_blocks(2);                           // allow two blocks to be voted on (the last two blocks of a round)
+   BOOST_REQUIRE_EQUAL(sb->producer, orig_producer);   // should still use orig producer
+
+   // now switch should happen in the next block, as it is the first block of a round
+   // -------------------------------------------------------------------------------
+   sb = A.produce_block();
+   BOOST_REQUIRE_NE(sb->producer, orig_producer);      // verify switch has happened
+   BOOST_REQUIRE_EQUAL(sb->timestamp.slot % prod_rep, 0); // verify first block of a round
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+//     Verify that a proposer policy becomes active when finality has advanced enough to make it pending
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(pending_proposer_policy_becomes_active_without_finality, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0];
+
+   auto sb = A.produce_block();
+   auto orig_producer = sb->producer;
+   auto orig_version = A.control->active_producers().version;
+
+   while (A.head().timestamp().slot % prod_rep >= prod_rep - 4)
+      A.produce_block();                  // make sure the next block is not one of the last three blocks of a round
+
+   const vector<account_name> producers { "pa"_n, "pb"_n };
+
+   A.create_accounts(producers);
+   A.tester::set_producers(producers);    // push the action to update the producer schedule
+   A.produce_block();                     // produce a block that will include the policy change transaction
+   A.produce_blocks(12);                  // produce 12 blocks which guarantees that the proposer policy is pending
+
+   // split network { A, B } and { C, D }
+   // Regardless of how many blocks A produces, finality will not advance
+   // by more than one (1 QC in flight)
+   // -------------------------------------------------------------------
+   const std::vector<size_t> partition {2, 3};
+   set_partition(partition);
+
+   sb = A.produce_block();                // produce one more block for lib final advance (in-flight QC)
+
+   BOOST_REQUIRE_EQUAL(sb->producer, orig_producer);
+   BOOST_REQUIRE_EQUAL(A.control->active_producers().version, orig_version);
+   BOOST_REQUIRE(!!A.control->pending_producers());
+   BOOST_REQUIRE_GT(A.control->pending_producers()->version, orig_version);
+   auto orig_lib_num = A.lib_number;
+
+   // produce `prod_rep` more blocks. Finality will not be advancing anymore, but still the new policy
+   // will become active because it was already pending.
```

Comment:
This is true, but it is also true that the policy will become active as long as it was simply _proposed_ prior to the last final block when finality stalled.

Reply:
@arhag I addressed your latest comments. Thanks for the thorough review!